### ISSUE TITLE
[Wave 04] LangGraph runtime + checkpoint/control realization for #69~#78

### ISSUE THREAD

**Issue by bonggyulim** [solar-ai-dev/mcp-work-agent#101]
## Fixed branch / dependency
Work only on `refactor/canonical-architecture-migration`.

Start only from the current pushed post-#100 inventory HEAD. Do not start #102 until this issue is complete, tested, committed, and pushed.

```text
#81 → #98 → #99 → #100 → #101 → #102 → #103 → #97
```

### Fixed #101 start revision

Start #101 only from:

```text
72ae3871d2b96d5d61c261988756210128875b11
```

This revision is exactly one commit after pushed #100 completion SHA `6ec3ff49a5f1e98afa5ff1b5a5ac4ff2fa9c5a3d`. The only change in `6ec3ff... → 72ae387...` is the addition of:

```text
implementation-inventory/domain-canonical-current-mapping-453e7f0c.md
```

No production code changed in that commit. Re-verify the actual branch HEAD before editing; if it differs, stop and report the mismatch.

### Phase-2 implementation inventory input

Use the checked-in mapping inventory:

```text
implementation-inventory/domain-canonical-current-mapping-453e7f0c.md
```

This file is **implementation inventory/evidence, not semantic or architecture authority**. Its exhaustive Domain model/transition investigation snapshot is `453e7f0c3fb5305775f709d91fe001673b5e0651`. #100 changed no `src/google_work_agent/domain/**` files, so the 54-row Domain model/transition dispositions remain usable as preservation-first evidence; however, Application/Port/caller evidence outside `domain/**` is living evidence and must be revalidated against the current #101 start HEAD before implementation.

If the mapping inventory conflicts with Canonical Source or actual current code, Canonical Source wins and the mapping must be treated as stale evidence, not as permission to invent a new design.

Do not interpret `PARTIAL` / `OWNER_PARTIAL` as permission for greenfield rewrite. Existing compatible behavior, tests, registries, checkpoint mechanics, and runtime wiring remain primary reuse candidates.

## Mission
Close the **concrete LangGraph / checkpoint / WorkflowExecution / registry / service-lifecycle realization** beneath the canonical #100 Application boundaries.

Do not reimplement Domain or Application semantics in adapters. Do not replace working runtime machinery merely to obtain canonical filenames.

Disposition preference:

```text
KEEP
→ MOVE / RENAME / MOVE_RENAME
→ SPLIT / MERGE
→ targeted runtime/crash-order correction
→ REWRITE only when preservation is crash-unsafe
→ DELETE only after binding/caller cut-over
→ CREATE only if no reusable realization exists
```

## Authority
Read synchronized `00 → 03 → 06 → 07 → 08 → 10 → 15 → 16/05,06,07,08,11,12,13 → 12/12-A → Ledger → completed #98~#100 → actual code/tests`.

06 owns graph/checkpoint/resume semantics. 07 owns typed Port/control contracts. 10 owns startup/shutdown order. 16 owns placement and single concrete authority.

## Verified current-code baseline — HEAD before #81
Baseline investigated: `2f344e4b93ace4d36a30fd0b6a8d6b12f1036c9c`. Re-verify after #100.

### A. NodeRegistry and ResumeTargetRegistry are already canonical KEEP authorities
Current:

- `adapters/langgraph/registry/node_registry.py → NodeRegistry`
- `adapters/langgraph/registry/resume_target_registry.py → ResumeTargetRegistry`

The investigated code already has:
- the 35 current Agent runtime-node mapping;
- profile→semantic-owner→compiled-subgraph binding;
- exact Main resume-stage set `RETRIEVAL_ENTRY | PLANNING_ENTRY | REVIEW_ENTRY | PREFLIGHT | READ_EXECUTION | VERIFICATION | RECOVERY | CANCEL_RESOLUTION`;
- `ACTION_EXECUTION` absent from resumable Main stages;
- fail-closed graph-version/owner/node/subgraph validation.

**Disposition: `KEEP`. Do not rebuild these registries or create a second stage/profile table.**

Also preserve the already-correct fact that `CheckpointEffectiveBindingResolver` in #100 delegates target legality to the injected registry. The old “private duplicate target table” finding is already closed at this baseline.

### B. BackgroundRunExecutor has strong persisted-admission/checkpoint mechanics but a one-shot control crash gap
Current:

- `adapters/langgraph/runtime/background_run_executor.py → BackgroundRunExecutorAdapter`

Reusable mechanics to KEEP:
- process-local queue is only an optimization over persisted admissions;
- exact persisted admission is reloaded before consumption;
- START/RESUME binding checks;
- execution-admission ID and active handoff lineage are written to checkpoint;
- checkpoint is flushed before handoff settlement;
- stale/mismatched admission does not immediately invoke owner I/O.

Confirmed gap:
- `_prepare_admission_checkpoint()` records admission/lineage/target, but does not durably materialize the handoff's typed one-shot control into the runnable native checkpoint/state;
- `_consume()` can then call `mark_consumed_and_clear_payload(...)` and commit before `_invoke_semantic_owner(...)`;
- after that settlement, a crash can leave the original Confirmation/ContextAdjustment control cleared without proof that its semantic effect exists in runnable checkpoint state.

**Required disposition: `KEEP + targeted ordering/materialization correction`.**

Required order for NORMAL one-shot control:

```text
reload exact persisted admission/handoff
→ validate current Run/binding/target
→ materialize admission + typed control into adapter-owned runnable checkpoint/native resume state
→ flush durable checkpoint
→ prove checkpoint contains the applied control lineage/evidence
→ settlement CAS / clear handoff payload
→ only if settled/current: semantic owner invocation
```

For `CONSUMED_CONTINUATION_RECOVERY`, original one-shot payload application count must be exactly 0; resume from the latest authorized descendant checkpoint only.

Do not move opaque checkpoint mutation into Application.

### C. Production composition already has useful canonical wiring — preserve it
Current `api/composition.py` already:
- binds one `BackgroundRunExecutorAdapter`;
- constructs one `ScheduleRunExecutionHandler`;
- injects `CheckpointEffectiveBindingResolver`;
- constructs and injects `RequireRecoveryHandler` into Redrive;
- constructs one `WorkflowHandoffReconciliationLoop`.

**KEEP this composition skeleton.** Do not re-add already existing bindings.

After #100 changes, update only the signatures/wiring needed for:
- durable one-shot control materialization;
- richer redrive durable-progress/has-more result;
- deterministic startup drain;
- exact shutdown ordering.

### D. WorkflowHandoffReconciliationLoop is already thin — keep it thin
Current:

- `adapters/system/workflow_handoff_reconciliation_loop.py`

It only wakes/calls `RedriveWorkflowHandoffsHandler` on a bounded interval and owns no Repository/WEP/LangGraph semantics.

**Disposition: `KEEP + minimal driver adaptation`.** If #100 adds `has_more`/durable-progress, use that only to drive bounded passes; do not move precedence or recovery decisions into the loop.

### E. FastAPI lifespan currently does not encode the canonical startup ordering
Current `api/lifespan.py`:
- schedules every `container.startup_callbacks` concurrently with `asyncio.ensure_future(...)`;
- immediately starts `container.local_run_coordinator`;
- only later yields READY/service operation;
- shutdown stops the LocalRunCoordinator and then generic callbacks.

This is not sufficient evidence for the synchronized canonical order:

```text
SQLite/Migration/Checkpoint readiness
→ Connector MCP + Tool/Schema readiness
→ LLM runtime load
→ ReconcileInflightExecutions bounded startup drain
→ RedriveWorkflowHandoffs bounded initial drain to no-progress/has_more closure
→ WorkflowHandoffReconciliationLoop start
→ READY
```

Required correction:
- preserve the lifespan structure but replace concurrent unordered startup with the exact ordered injected lifecycle sequence;
- prove the #100 Redrive handler performs the initial drain before the live loop starts;
- prove shutdown stops/drains the same live loop before lower-level runtime teardown;
- no legacy open-Run scan may substitute for durable handoff reconciliation.

### F. LocalRunCoordinator is a broad legacy runtime authority and must be split/cut over, not blindly deleted
Current `application/coordinator.py → LocalRunCoordinator` still directly owns a process-local queue and calls `WorkflowRuntime.resume`, `request_cancel`, `recover_open_run`, outcome/finalization logic, and private QueryService/UoW collaborators.

Current API lifespan starts this coordinator directly. Current routes/ResumeRun also still reference it before #100 cut-over.

After #100 caller migration:
- identify any remaining genuinely unmigrated responsibility;
- move reusable low-level queue/event mechanics only if a current canonical owner actually needs them;
- remove START/resume/recovery/cancel semantic execution responsibilities that overlap canonical Application + WorkflowExecutionPort paths;
- delete the broad coordinator only when all its live responsibilities have either moved or are proven obsolete.

Do not translate the whole coordinator into another generic runtime manager.

### G. ApiContainer currently performs concrete fallback construction outside composition
Current `api/container.py` imports `NodeRegistry`/`ResumeTargetRegistry` and constructs a registry in `__post_init__` when none is supplied. It also constructs selection-handle issuer/resolver fallback objects.

The synchronized Repository Architecture requires concrete production construction to be centralized in `api/composition.py` except registered exceptions.

Required disposition:
- preserve the dependency-record/container role;
- MOVE concrete construction into the composition root;
- keep container fields as already-built abstract/canonical dependencies;
- remove fallback construction only after all tests/composition callers supply the canonical dependencies.

Do not delete and recreate the registry implementations themselves.

## Required runtime/crash tests
Use real persistence/checkpoint seams where possible:

1. admission claim commit → crash before worker consumption;
2. control checkpoint commit → crash before settlement;
3. settlement after control durability → crash before owner invocation;
4. restart after one descendant and after 2+ descendants;
5. historical CONSUMED H1 plus current active-lineage H2;
6. current Run authority changes after admission and before settlement;
7. stale admission/target → owner Node/LLM/Connector I/O = 0;
8. normal Confirmation control applied exactly once before payload clear;
9. consumed-continuation original payload reinjection = 0;
10. unknown/stale owner/node/stage/version fails through the existing canonical registries;
11. ordered startup drains execute exactly once before live loop/READY;
12. shutdown stops/drains live loop exactly once;
13. committed PENDING truth recovers without browser replay or legacy open-Run scan;
14. LocalRunCoordinator no longer owns a migrated runtime semantic path;
15. behavior preservation for retained checkpoint/registry/executor mechanics.

## Done condition
At one pushed revision:

- WorkflowExecutionPort production binding = 1;
- existing NodeRegistry/ResumeTargetRegistry remain the sole legality authorities;
- one-shot control lost-decision crash cut = 0;
- consumed-continuation reinjection = 0;
- stale owner I/O paths = 0;
- exact active lineage survives crash/restart;
- startup order is deterministic and canonical;
- live-loop semantic authority remains only RedriveWorkflowHandoffsHandler;
- migrated LocalRunCoordinator execution/resume/recovery bypasses = 0;
- concrete production construction outside composition = 0 except registered exceptions;
- Application opaque checkpoint mutation = 0;
- unnecessary runtime/test loss = 0;
- unjustified `DELETE + CREATE` churn = 0;
- targeted runtime/checkpoint/crash/integration/architecture suites PASS.

## Completion report
Report runtime responsibility/binding/caller inventory, dispositions, preserved mechanics, exact changed ordering, removed legacy coordinator responsibilities and caller proof, crash matrix, every rewrite/delete/create justification, exact tests/results, and pushed SHA.

**Comment by bonggyulim** [comment]
## 중간 체크포인트 (2026-08-27)

요청에 따라 이 지점에서 구현을 중단하고 현재 상태를 기록합니다. **Issue #101은 아직 완료가 아닙니다.**

### 기준/브랜치
- 브랜치: `refactor/canonical-architecture-migration`
- 검증한 시작 SHA: `72ae3871d2b96d5d61c261988756210128875b11`
- 현재 HEAD: 동일 (아직 commit/push 없음)
- #102 작업 시작 안 함

### 현재까지 구현
- 기존 `NodeRegistry` / `ResumeTargetRegistry` 유지
- `BackgroundRunExecutorAdapter`의 persisted admission → typed control native checkpoint materialization → flush/evidence proof → settlement → semantic owner invoke 순서 반영
- Confirmation은 native `__resume__`, ContextAdjustment/Retrieval restart는 native state update/goto로 materialize
- `CONSUMED_CONTINUATION_RECOVERY`에서 원본 one-shot control을 재주입하지 않고 빈 recovery payload로 owner 호출
- native checkpoint의 latest active descendant를 resume target으로 projection하는 `NativeCheckpointTargetResolver` 추가
- deterministic startup 순서: inflight execution reconciliation drain → WorkflowHandoff redrive drain → live loop
- shutdown 순서에서 live loop 정지 후 worker/checkpoint/lower resource 정리 유지
- `ReconcileInflightExecutionsHandler` 및 SQLite reconciliation candidate query 추가
- POST_BEGIN orphan을 UNKNOWN_RESULT + RecoveryContext로 같은 UoW에서 전환하도록 보강
- approve / modify / cancel / generic resume production caller를 durable WorkflowHandoff staging + scheduling으로 cut-over
- `ApiContainer`의 registry/selection-handle concrete fallback 제거, launcher composition root로 이동
- `LocalRunCoordinator` production caller 0 확인 후 source 및 coordinator 전용 unit test 삭제
- `READ_EXECUTION`, `CANCEL_RESOLUTION`의 native action-execution route realization 추가

### 현재 검증
- 집중 unit/integration 회귀: **45 passed**
  - API lifespan/container
  - deferred startup
  - approve/modify/resume handler
  - BackgroundRunExecutor
  - external control handoff crash recovery
- `git diff --check`: 통과
- real LangGraph 수동 검증:
  - Confirmation pending resume 1회 적용
  - ContextAdjustment durable typed state + goto 적용

### 아직 남은 작업 / gate
- Issue 본문의 crash matrix 1~15 전체를 명시적 자동 테스트로 완성
- `ReconcileInflightExecutions` 4 candidate kind 및 connector I/O=0 전용 테스트 추가
- active lineage(H1/H2, 2+ descendants, stale authority I/O=0) 확대 검증
- component/integration 전체와 full unit regression 실행
- architecture/static gate 및 targeted Ruff 재실행
- formatter가 만든 관련 없는 line-ending-only dirty 표시는 semantic diff와 분리해 최종 정리 필요
- 최종 caller/binding census 및 duplicate authority negative proof
- closure 후에만 commit/push, local=remote/clean worktree 확인

현재 변경은 working tree에만 있으며 **커밋과 푸시는 의도적으로 하지 않았습니다.**

**Comment by bonggyulim** [comment]
Execution-order reset: this Wave is no longer the active work authority. The pushed checkpoint commit `93f03a918cbd9cfd047da1c1b1ee70aca76da8f6` is retained as implementation evidence, but full #101 closure was not proven. Current executable authority is #104, which returns to full Canonical Domain closure first. Remaining runtime validation/corrections must be re-evaluated only after Domain is COMPLETE.
